pkg/servicescm/servicescm.go

```diff
@@ -125,6 +134,150 @@ func Generate(name, namespace string) (*core.ConfigMap, error) {
 	return servicesConfigMap, nil
 }
 
+// Parse converts ConfigMap data into the objects representing a Windows services ConfigMap schema
+// Returns error if the given data is invalid in structure or contents.
+func Parse(dataFromCM map[string]string) (*data, error) {
+	if len(dataFromCM) != 2 {
+		return nil, errors.New("services ConfigMap should have exactly 2 keys")
+	}
+
+	services := &[]Service{}
+	files := &[]FileInfo{}
+
+	value, ok := dataFromCM[servicesKey]
+	if !ok {
+		return nil, errors.Errorf("expected key %s does not exist", servicesKey)
+	}
+	if err := json.Unmarshal([]byte(value), services); err != nil {
+		return nil, err
+	}
+
+	value, ok = dataFromCM[filesKey]
+	if !ok {
+		return nil, errors.Errorf("expected key %s does not exist", filesKey)
+	}
+	if err := json.Unmarshal([]byte(value), files); err != nil {
+		return nil, err
+	}
+
+	return newData(services, files)
+}
+
+// validate ensures the given object represents a valid services ConfigMap, made up of only required services and files.
+// The validation also ensures bootstrap services are defined to always start before controller services.
+func (cmData *data) validate() error {
+	if err := validateRequiredContent(cmData.Services, cmData.Files); err != nil {
+		return err
+	}
+	if err := validateDependencies(cmData.Services); err != nil {
+		return err
+	}
+	return validatePriorities(cmData.Services)
+}
+
+func validateRequiredContent(services []Service, files []FileInfo) error {
+	// Validate services
+	if len(services) != len(*requiredServices) {
+		return errors.New("Unexpected number of services")
+	}
+	for _, requiredSvc := range *requiredServices {
+		// isPresent checks if the required service exists within the given services slice
+		isPresent := func() error {
+			for _, svc := range services {
+				if reflect.DeepEqual(requiredSvc, svc) {
+					return nil
+				}
+			}
+			return errors.Errorf("Required service %s is not present with expected configuration", requiredSvc.Name)
+		}
+		if err := isPresent(); err != nil {
+			return err
+		}
+	}
+
+	// Validate files
+	if len(files) != len(*requiredFiles) {
+		return errors.New("Unexpected number of files")
+	}
+	for _, requiredFile := range *requiredFiles {
+		// isPresent checks if the required file exists within the given files slice
+		isPresent := func() error {
+			for _, file := range files {
+				if reflect.DeepEqual(requiredFile, file) {
+					return nil
+				}
+			}
+			return errors.Errorf("Required file %s is not present as expected", requiredFile.Path)
```

Comment:
Same comment here.

PS: with Go 1.18, you could use a generics and make a common method across files and services.

Reply:
We've not yet moved to 1.18 so I will leave a TODO to consolidate the helpers when we do.